File: apps/android/app/src/main/java/com/rsr41/oracle/data/repository/SettingsRepositoryImpl.kt
package com.rsr41.oracle.data.repository

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.*
import com.rsr41.oracle.domain.model.AppLanguage
import com.rsr41.oracle.domain.model.CalendarType
import com.rsr41.oracle.domain.model.ThemeMode
import com.rsr41.oracle.repository.SettingsRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class SettingsRepositoryImpl @Inject constructor(
    private val dataStore: DataStore<Preferences>
) : SettingsRepository {

    private object PreferencesKeys {
        val THEME_MODE = stringPreferencesKey("theme_mode")
        val APP_LANGUAGE = stringPreferencesKey("app_language")
        val DEFAULT_CALENDAR_TYPE = stringPreferencesKey("default_calendar_type")
        val LUNAR_LEAP_MONTH_ENABLED = booleanPreferencesKey("lunar_leap_month_enabled")
    }

    override val themeMode: Flow<ThemeMode> = dataStore.data
        .map { preferences ->
            val modeName = preferences[PreferencesKeys.THEME_MODE] ?: ThemeMode.SYSTEM.name
            try {
                ThemeMode.valueOf(modeName)
            } catch (e: Exception) {
                ThemeMode.SYSTEM
            }
        }

    override val appLanguage: Flow<AppLanguage> = dataStore.data
        .map { preferences ->
            val langName = preferences[PreferencesKeys.APP_LANGUAGE] ?: AppLanguage.SYSTEM.name
            try {
                AppLanguage.valueOf(langName)
            } catch (e: Exception) {
                AppLanguage.SYSTEM
            }
        }

    override val defaultCalendarType: Flow<CalendarType> = dataStore.data
        .map { preferences ->
            val typeName = preferences[PreferencesKeys.DEFAULT_CALENDAR_TYPE] ?: CalendarType.SOLAR.name
            try {
                CalendarType.valueOf(typeName)
            } catch (e: Exception) {
                CalendarType.SOLAR
            }
        }

    override val isLunarLeapMonthEnabled: Flow<Boolean> = dataStore.data
        .map { preferences ->
            preferences[PreferencesKeys.LUNAR_LEAP_MONTH_ENABLED] ?: false
        }

    override suspend fun setThemeMode(mode: ThemeMode) {
        dataStore.edit { preferences ->
            preferences[PreferencesKeys.THEME_MODE] = mode.name
        }
    }

    override suspend fun setAppLanguage(language: AppLanguage) {
        dataStore.edit { preferences ->
            preferences[PreferencesKeys.APP_LANGUAGE] = language.name
        }
    }

    override suspend fun setDefaultCalendarType(type: CalendarType) {
        dataStore.edit { preferences ->
            preferences[PreferencesKeys.DEFAULT_CALENDAR_TYPE] = type.name
        }
    }

    override suspend fun setLunarLeapMonthEnabled(enabled: Boolean) {
        dataStore.edit { preferences ->
            preferences[PreferencesKeys.LUNAR_LEAP_MONTH_ENABLED] = enabled
        }
    }
}

File: apps/ios/OracleIOS/OracleIOS/OracleIOS_Sources/Data/Repositories/SettingsRepositoryImpl.swift
import Foundation

/// 설정 Repository 구현체
/// UserDefaults 기반 (Android DataStore 대응)
final class SettingsRepositoryImpl: SettingsRepository {
    private let userDefaults: UserDefaults
    
    // Keys
    private enum Keys {
        static let calendarType = "defaultCalendarType"
        static let themeMode = "themeMode"
        static let appLanguage = "appLanguage"
        static let faceConsent = "faceConsent"
    }
    
    init(userDefaults: UserDefaults = .standard) {
        self.userDefaults = userDefaults
    }
    
    // MARK: - Calendar Type
    
    var defaultCalendarType: CalendarType {
        get async throws {
            guard let value = userDefaults.string(forKey: Keys.calendarType),
                  let type = CalendarType(rawValue: value) else {
                return .solar
            }
            return type
        }
    }
    
    func setDefaultCalendarType(_ type: CalendarType) async throws {
        userDefaults.set(type.rawValue, forKey: Keys.calendarType)
    }
    
    // MARK: - Theme Mode
    
    var themeMode: ThemeMode {
        get async throws {
            guard let value = userDefaults.string(forKey: Keys.themeMode),
                  let mode = ThemeMode(rawValue: value) else {
                return .system
            }
            return mode
        }
    }
    
    func setThemeMode(_ mode: ThemeMode) async throws {
        userDefaults.set(mode.rawValue, forKey: Keys.themeMode)
    }
    
    // MARK: - App Language
    
    var appLanguage: AppLanguage {
        get async throws {
            guard let value = userDefaults.string(forKey: Keys.appLanguage),
                  let lang = AppLanguage(rawValue: value) else {
                return .system
            }
            return lang
        }
    }
    
    func setAppLanguage(_ language: AppLanguage) async throws {
        userDefaults.set(language.rawValue, forKey: Keys.appLanguage)
    }
    
    // MARK: - Face Consent
    
    var faceConsent: Bool {
        get async throws {
            userDefaults.bool(forKey: Keys.faceConsent)
        }
    }
    
    func setFaceConsent(_ consented: Bool) async throws {
        userDefaults.set(consented, forKey: Keys.faceConsent)
    }
}
